### PR TITLE
Fix and make TaintManager harder to break before we move it out of NC

### LINKED ISSUE #45342: Taint based evictions is broken
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

**Kubernetes version** (use `kubectl version`):
HEAD

```
root@k8s001:~/cases/taint# kubectl version
Client Version: version.Info{Major:"1", Minor:"7+", GitVersion:"v1.7.0-alpha.2.718+3b8aac9b22f862", GitCommit:"3b8aac9b22f86288ebc40073c8705bd9d4484c33", GitTreeState:"clean", BuildDate:"2017-05-04T09:14:20Z", GoVersion:"go1.8.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"7+", GitVersion:"v1.7.0-alpha.2.718+3b8aac9b22f862", GitCommit:"3b8aac9b22f86288ebc40073c8705bd9d4484c33", GitTreeState:"clean", BuildDate:"2017-05-04T09:14:20Z", GoVersion:"go1.8.1", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**:
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**:
```
root@k8s001:~/cases/taint# kubectl get nodes -owide
NAME      STATUS    AGE       VERSION                             EXTERNAL-IP   OS-IMAGE           KERNEL-VERSION
k8s001    Ready     28d       v1.7.0-alpha.2.718+3b8aac9b22f862   <none>        Ubuntu 16.04 LTS   4.4.0-72-generic
k8s004    Ready     28d       v1.7.0-alpha.2.718+3b8aac9b22f862   <none>        Ubuntu 16.04 LTS   4.4.0-75-generic
```

**What happened**:
1) Enable `TaintBasedEvictions`

```
/opt/bin/kube-controller-manager --logtostderr=false --log-dir=/var/log/kubernetes --v=4 --master=http://192.168.56.11:8080 --cluster-name=kubernetes --root-ca-file=/srv/kubernetes/ca.crt --service-account-private-key-file=/srv/kubernetes/server.key --feature-gates=TaintBasedEvictions=true
```

2) Create a pod.

pod yaml
```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  labels:
    run: my-nginx
  name: my-nginx
  namespace: default
spec:
  replicas: 1
  selector:
    matchLabels:
      run: my-nginx
  template:
    metadata:
      labels:
        run: my-nginx
    spec:
      containers:
      - image: nginx:1.8.1
        imagePullPolicy: IfNotPresent
        name: my-nginx
        ports:
        - containerPort: 80
          protocol: TCP
```

After deploy, it is running on `k8s004`
```
root@k8s001:~/cases/taint# kubectl get pods -owide
NAME                        READY     STATUS    RESTARTS   AGE       IP                NODE
my-nginx-1257052682-lkjpq   1/1       Running   0          6m        192.168.124.147   k8s004
```

3) Add taint to `k8s004`

```
kubectl taint nodes k8s004 x=y:NoExecute
```

```
root@k8s001:~/cases/taint# kubectl describe nodes k8s004
Name:			k8s004
Role:
Labels:			beta.kubernetes.io/arch=amd64
			beta.kubernetes.io/os=linux
			kubernetes.io/hostname=k8s004
			role=ok
Annotations:		node.alpha.kubernetes.io/ttl=0
			volumes.kubernetes.io/controller-managed-attach-detach=true
Taints:			x=y:NoExecute
CreationTimestamp:	Thu, 06 Apr 2017 10:18:10 +0800
Conditions:
  Type			Status	LastHeartbeatTime			LastTransitionTime			Reason				Message
  ----			------	-----------------			------------------			------				-------
  OutOfDisk 		False 	Thu, 04 May 2017 17:56:12 +0800 	Thu, 04 May 2017 17:03:14 +0800 	KubeletHasSufficientDisk 	kubelet has sufficient disk space available
  MemoryPressure 	False 	Thu, 04 May 2017 17:56:12 +0800 	Thu, 04 May 2017 17:03:14 +0800 	KubeletHasSufficientMemory 	kubelet has sufficient memory available
  DiskPressure 		False 	Thu, 04 May 2017 17:56:12 +0800 	Thu, 04 May 2017 17:03:14 +0800 	KubeletHasNoDiskPressure 	kubelet has no disk pressure
  Ready 		True 	Thu, 04 May 2017 17:56:12 +0800 	Thu, 04 May 2017 17:26:50 +0800 	KubeletReady 			kubelet is posting ready status. AppArmor enabled
Addresses:
  LegacyHostIP:	192.168.56.14
  InternalIP:	192.168.56.14
  Hostname:	k8s004
Capacity:
 cpu:		4
 memory:	4046240Ki
 pods:		110
Allocatable:
 cpu:		4
 memory:	3943840Ki
 pods:		110
System Info:
 Machine ID:			1ed0241c581fcc54296c57315864e03e
 System UUID:			AB6794AC-B3E9-4E95-80CB-9C7E64D2AB09
 Boot ID:			3dd022be-280b-4018-9929-8a9900ca359a
 Kernel Version:		4.4.0-75-generic
 OS Image:			Ubuntu 16.04 LTS
 Operating System:		linux
 Architecture:			amd64
 Container Runtime Version:	docker://1.12.5
 Kubelet Version:		v1.7.0-alpha.2.718+3b8aac9b22f862
 Kube-Proxy Version:		v1.7.0-alpha.2.718+3b8aac9b22f862
ExternalID:			k8s004
Non-terminated Pods:		(1 in total)
  Namespace			Name					CPU Requests	CPU Limits	Memory Requests	Memory Limits
  ---------			----					------------	----------	---------------	-------------
  default			my-nginx-1257052682-lkjpq		0 (0%)		0 (0%)		0 (0%)		0 (0%)
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  CPU Requests	CPU Limits	Memory Requests	Memory Limits
  ------------	----------	---------------	-------------
  0 (0%)	0 (0%)		0 (0%)		0 (0%)
Events:
  FirstSeen	LastSeen	Count	From			SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----			-------------	--------	------			-------
  52m		52m		1	kubelet, k8s004				Normal		Starting		Starting kubelet.
  52m		52m		1	kubelet, k8s004				Normal		NodeAllocatableEnforced	Updated Node Allocatable limit across pods
  52m		52m		2	kubelet, k8s004				Normal		NodeHasSufficientDisk	Node k8s004 status is now: NodeHasSufficientDisk
  52m		52m		2	kubelet, k8s004				Normal		NodeHasSufficientMemory	Node k8s004 status is now: NodeHasSufficientMemory
  52m		52m		2	kubelet, k8s004				Normal		NodeHasNoDiskPressure	Node k8s004 status is now: NodeHasNoDiskPressure
  52m		52m		1	kubelet, k8s004				Normal		NodeNotReady		Node k8s004 status is now: NodeNotReady
  52m		52m		1	kubelet, k8s004				Normal		NodeReady		Node k8s004 status is now: NodeReady
  29m		29m		1	kube-proxy, k8s004			Normal		Starting		Starting kube-proxy.
  29m		29m		1	kubelet, k8s004				Normal		Starting		Starting kubelet.
  29m		29m		1	kubelet, k8s004				Normal		NodeAllocatableEnforced	Updated Node Allocatable limit across pods
  29m		29m		1	kubelet, k8s004				Normal		NodeHasSufficientDisk	Node k8s004 status is now: NodeHasSufficientDisk
  29m		29m		1	kubelet, k8s004				Normal		NodeHasSufficientMemory	Node k8s004 status is now: NodeHasSufficientMemory
  29m		29m		1	kubelet, k8s004				Normal		NodeHasNoDiskPressure	Node k8s004 status is now: NodeHasNoDiskPressure
  29m		29m		1	kubelet, k8s004				Normal		NodeNotReady		Node k8s004 status is now: NodeNotReady
  29m		29m		1	kubelet, k8s004				Normal		NodeReady		Node k8s004 status is now: NodeReady
```

4) The pod running on `k8s004` does not get evicted.

```
root@k8s001:~/cases/taint# kubectl get pods -owide
NAME                        READY     STATUS    RESTARTS   AGE       IP                NODE
my-nginx-1257052682-lkjpq   1/1       Running   0          8m        192.168.124.147   k8s004
root@k8s001:~/cases/taint# kubectl describe pods
Name:		my-nginx-1257052682-lkjpq
Namespace:	default
Node:		k8s004/192.168.56.14
Start Time:	Thu, 04 May 2017 17:48:29 +0800
Labels:		pod-template-hash=1257052682
		run=my-nginx
Annotations:	kubernetes.io/created-by={"kind":"SerializedReference","apiVersion":"v1","reference":{"kind":"ReplicaSet","namespace":"default","name":"my-nginx-1257052682","uid":"d3d8c316-30ae-11e7-993a-08002759511a...
Status:		Running
IP:		192.168.124.147
Created By:	ReplicaSet/my-nginx-1257052682
Controlled By:	ReplicaSet/my-nginx-1257052682
Containers:
  my-nginx:
    Container ID:	docker://130537e470866308223620786b7abc266b24eb215c9211562d9ef98c5232294a
    Image:		nginx:1.8.1
    Image ID:		docker-pullable://nginx@sha256:9b3e9f189890ef9d6713c3384da3809731bdb0bff84e7b68da330ebadf533085
    Port:		80/TCP
    State:		Running
      Started:		Thu, 04 May 2017 17:48:30 +0800
    Ready:		True
    Restart Count:	0
    Environment:	<none>
    Mounts:
      /var/run/secrets/kubernetes.io/serviceaccount from default-token-0qlw6 (ro)
Conditions:
  Type		Status
  Initialized 	True
  Ready 	True
  PodScheduled 	True
Volumes:
  default-token-0qlw6:
    Type:	Secret (a volume populated by a Secret)
    SecretName:	default-token-0qlw6
    Optional:	false
QoS Class:	BestEffort
Node-Selectors:	<none>
Tolerations:	<none>
Events:
  FirstSeen	LastSeen	Count	From			SubObjectPath			Type		Reason		Message
  ---------	--------	-----	----			-------------			--------	------		-------
  8m		8m		1	default-scheduler					Normal		Scheduled	Successfully assigned my-nginx-1257052682-lkjpq to k8s004
  8m		8m		1	kubelet, k8s004		spec.containers{my-nginx}	Normal		Pulled		Container image "nginx:1.8.1" already present on machine
  8m		8m		1	kubelet, k8s004		spec.containers{my-nginx}	Normal		Created		Created container with id 130537e470866308223620786b7abc266b24eb215c9211562d9ef98c5232294a
  8m		8m		1	kubelet, k8s004		spec.containers{my-nginx}	Normal		Started		Started container with id 130537e470866308223620786b7abc266b24eb215c9211562d9ef98c5232294a
root@k8s001:~/cases/taint#
```

I also test another automatic case by killing `kubelet` on `k8s004`, and I can see the `controller-manager` add a taint `node.alpha.kubernetes.io/unreachable:NoExecute`, but the pods also does not get evicted.

**What you expected to happen**:
Pods should be evicted.

**How to reproduce it** (as minimally and precisely as possible):

**Anything else we need to know**:

### PR DESCRIPTION
**Pull request by gmarek**
Fix #45342

cc @gyliu513

### CODE REVIEW

File: pkg/controller/node/nodecontroller.go [hunk 1]
```diff
@@ -314,6 +314,9 @@ func NewNodeController(
 	})
 	nc.podInformerSynced = podInformer.Informer().HasSynced
 
+	addNodeFunctions := []func(obj *v1.Node) error{}
```
Comment (wojtek-t) [inline pkg/controller/node/nodecontroller.go]: As discussed offline, we should have multiple "AddEventHandlers" functions instead of this.

**Comment by gmarek** [comment]
@k8s-bot bazel test this

**Comment by gmarek** [comment]
@k8s-bot bazel test this

**Comment by gyliu513** [comment]
Thanks @gmarek !

**Pull request merged**